### PR TITLE
Flyweights

### PR DESCRIPTION
**Pull request by nezaniel**
Makes some ESCR objects flyweights to improve performance and memory footprint

### CODE REVIEW

**Comment by nezaniel** [comment]
FYI: This is based on https://github.com/neos/neos-development-collection/pull/3592, which should be merged first

File: Neos.ContentRepository/Classes/Domain/ContentStream/ContentStreamIdentifier.php [hunk 1]
```diff
@@ -12,36 +10,44 @@
...
-    private $value;
+    private static array $instances = [];
+
+    private function __construct(
+        private string $value
+    ) {
+    }
 
-    private function __construct(string $value)
+    public static function instance(string $value): self
```
Comment (bwaidelich) [inline Neos.ContentRepository/Classes/Domain/ContentStream/ContentStreamIdentifier.php:38]: I slightly prefer "constant" over "instance" (or something that communicates that this doesn't necessarily create a new instance). But probably that's just a matter of getting used to.

However, this method should be private IMO:
```suggestion
    private static function instance(string $value): self
```

File: Neos.ContentRepository/Classes/Domain/ContentStream/ContentStreamIdentifier.php [hunk 2]
```diff
@@ -12,36 +10,44 @@
...
     /**
-     * @var string
+     * @var array<string,self>
      */
-    private $value;
+    private static array $instances = [];
+
+    private function __construct(
+        private string $value
+    ) {
```
Comment (bwaidelich) [inline Neos.ContentRepository/Classes/Domain/ContentStream/ContentStreamIdentifier.php:35]: nitpick: Why 3 lines for the constructor signature?

Comment (nezaniel) [reply Neos.ContentRepository/Classes/Domain/ContentStream/ContentStreamIdentifier.php:35]: With the constructor being the only place left where a property is declared, I prefer a separate line for that property to have all its attributes (type, readonly, scope) readily available.

File: Neos.ContentRepository/Classes/Domain/ContentStream/ContentStreamIdentifier.php (hunk 1 above)
Comment (nezaniel) [reply Neos.ContentRepository/Classes/Domain/ContentStream/ContentStreamIdentifier.php:38]: Already thought the same regarding instance() being private, especially as it is currently identical to fromString and could lead to confusion which one to use. I'd go with private, too.

File: Neos.ContentRepository/Classes/Domain/NodeType/NodeTypeName.php [hunk 3]
```diff
@@ -21,45 +22,49 @@
...
+    private function __construct(
+        private string $value
+    ) {
         if ($value === '') {
             throw new \InvalidArgumentException('Node type name must not be empty.', 1505835958);
         }
+    }
 
-        $this->value = $value;
+    public static function instance(string $value): self
```
Comment (bwaidelich) [inline Neos.ContentRepository/Classes/Domain/NodeType/NodeTypeName.php:42]:
```suggestion
    private static function instance(string $value): self
```

File: Neos.ContentRepository/Classes/Domain/ContentStream/ContentStreamIdentifier.php (hunk 2 above)
Comment (mhsdesign) [reply Neos.ContentRepository/Classes/Domain/ContentStream/ContentStreamIdentifier.php:35]: makes sense 👍

**Review by mhsdesign — APPROVED** [review]

File: Neos.ContentRepository/Classes/Domain/ContentStream/ContentStreamIdentifier.php (hunk 2 above)
Comment (bwaidelich) [reply Neos.ContentRepository/Classes/Domain/ContentStream/ContentStreamIdentifier.php:35]: Yes right, this makes sense!

**Review by bwaidelich — APPROVED** [review]
+1 by reading

**Pull request merged**